### PR TITLE
Add new `SFTDataset` to `megatron.core.datasets`

### PR DESCRIPTION
**Pull request by asolergi-nv**
- [X] I, the PR author, have personally reviewed every line of this PR.

## Summary

This PR replaces the JSONL-backed `SFTDataset` in `megatron/training/datasets/sft_dataset.py` with a new pre-tokenized, packed implementation under `megatron/core/datasets/sft_dataset.py`. The new dataset:

- Reads pre-tokenized chat conversations from Megatron's file prefixes (`.bin` / `.idx`) like pretraining. It's up to the user to apply the chat template and later on pretokenize the data with the `tools/preprocess_data.py` script.
- Packs variable-length conversations into fixed-capacity samples using the **Modified First-Fit Decreasing (MFFD)** bin-packing algorithm and caches the resulting indices on disk.
- Builds per-token **loss masks** by parsing each conversation against a configurable chat template (`ChatTemplateConfig` — Nemotron-3 default). Three flags control what counts toward the loss:
  - `train_on_assistant_responses_only`
  - `train_on_thinking_traces` (mask `<think>…</think>` regions)
  - `train_on_tool_calls` (mask `<tool_call>…</tool_call>` regions). Tool responses (environment output inside user turns) are always masked.
- Emits THD-packed outputs (`tokens`, `labels`, `loss_mask`, `cu_seqlens`) ready for Flash Attention varlen kernels.

A new helper, `preprocess_sft_batch`, is added to `megatron/core/utils.py` and called from both `pretrain_gpt.py` and `pretrain_hybrid.py`. It handles the runtime steps that used to happen inside the old `__getitem__`: padding each sub-sequence to a CP-divisible length, padding or truncating the packed sample to `seq_length`, and producing `position_ids`, `cu_seqlens_padded`, and `max_seqlen`.

A standalone debug script, `inspect_sft_file_prefix.py`, is added at the repo root for inspecting a `.bin`/`.idx` pair against the Nemotron-3 template. Will remove before merge.

User-facing reference docs for the dataset live in `megatron/core/datasets/sft.MD`.

---

### New: dataset and helpers

| File | What it does |
|------|--------------|
| `megatron/core/datasets/sft_dataset.py` *(new, ~830 lines)* | `SFTDataset`, `SFTDatasetConfig`, `ChatTemplateConfig`, `Nemotron3ChatTemplateConfig`, the MFFD packer (`pack_samples` + `_classify_items`), and the chat-template segment parser (`extract_segments`, `_split_tool_calls`, `find_subsequence`). |
| `megatron/core/datasets/sft.MD` *(new)* | Reference documentation: output schema, index-building, MFFD phases, segment roles, and worked examples for chat-only and tool-use trajectories. |
| `megatron/core/utils.py` *(+289 lines)* | Three SFT helpers: `pad_thd_sequences_for_cp` (round each segment up to a CP-divisible length), `pad_or_truncate_thd_tensors` (force the packed sample to exactly `max_seq_len`), and `preprocess_sft_batch` (TP-rank-0 batch preprocessing wrapper). |
| `inspect_sft_file_prefix.py` *(new, 332 lines)* | Standalone debug script that loads a `.bin`/`.idx` pair and prints per-segment role assignments using the hard-coded Nemotron-3 chat template. Not imported by training code. |

### Removed: the old SFTDataset

| File | Change |
|------|--------|
| `megatron/training/datasets/sft_dataset.py` | **Deleted.** The JSONL-backed `SFTDataset` / `SFTLowLevelDataset` are superseded by the pre-tokenized core implementation. |

### Training entrypoints wired to the new dataset

| File | Change |
|------|--------|
| `pretrain_gpt.py` | Import `SFTDataset`, `SFTDatasetConfig`, `IGNORE_INDEX` from `megatron.core.datasets.sft_dataset`. Call `preprocess_sft_batch` inside `get_batch` when `args.sft` is set. Return `SFTDatasetConfig` from `core_gpt_dataset_config_from_args` when in SFT mode. |
| `pretrain_hybrid.py` | Same wiring as `pretrain_gpt.py`. Also refactored `core_gpt_dataset_config_from_args` to build a `data_args` dict so the SFT and non-SFT configs share construction. |
| `megatron/elastification/pretrain_hybrid_flex.py` | Update `SFTDataset` import to point at the new module. |

### Tokenizer `add_special_tokens` plumbing

The dataset tokenizes chat-template delimiter strings (`<|im_start|>system\n`,
`<|im_end|>\n`, `<think>`, …) and must do so **without** the tokenizer
auto-inserting BOS/EOS. A new `add_special_tokens: bool = True` keyword is
added through the tokenizer stack:

| File | Change |
|------|--------|
| `megatron/core/tokenizers/text/libraries/abstract_tokenizer.py` | Add `add_special_tokens: bool = True` to the `text_to_ids` ABC signature. |
| `megatron/core/tokenizers/text/libraries/bytelevel_tokenizer.py` | Accept and ignore `add_special_tokens`. |
| `megatron/core/tokenizers/text/libraries/huggingface_tokenizer.py` | Accept `add_special_tokens` (caller-side responsibility — HF path is unchanged when `True`). |
| `megatron/core/tokenizers/text/libraries/sentencepiece_tokenizer.py` | Accept and ignore `add_special_tokens`. |
| `megatron/core/tokenizers/text/libraries/tiktoken_tokenizer.py` | Accept and ignore `add_special_tokens`. |
| `megatron/core/tokenizers/text/libraries/sft_tokenizer.py` | Forward `add_special_tokens` to `self._tokenizer.encode(...)`; tighten an `assert` with a message. |
| `megatron/core/tokenizers/text/libraries/null_tokenizer.py` | Accept `add_special_tokens` **and** change `text_to_ids` to a character-level encoding `[ord(c) % vocab_size for c in text]`. This makes the null tokenizer usable in the SFT unit tests (it can now tokenize arbitrary template strings reproducibly). |
| `megatron/core/tokenizers/text/text_tokenizer.py` | Plumb `add_special_tokens` through `tokenize(...)`. |
| `megatron/core/tokenizers/utils/build_tokenizer.py` | Tiny refactor: lift the `{NullTokenizer, NullMultimodalTokenizer} -> library name` mapping into a `NULL_TOKENIZERS` dict. |

---

### New tests

- **`tests/unit_tests/data/test_sft_dataset.py`** *(399 lines)* — end-to-end
  test of `SFTDataset` via `BlendedMegatronDatasetBuilder`.
  - Builds synthetic `.bin`/`.idx` files containing 4 conversation shapes
    (`simple`, `with_thinking`, `with_tool_calls`, `with_thinking_and_tool_calls`),
    with and without system turns.
  - Uses a `TestChatTemplateConfig` whose delimiter strings tokenize cleanly
    under the (now character-level) `NullTokenizer`.
  - Parametrized across `vocab_size ∈ {131072, 20000}` and all five
    `(train_on_assistant_responses_only, train_on_thinking_traces, train_on_tool_calls)`
    combinations allowed by the asserts in `SFTDatasetConfig.__post_init__`.
  - Verifies dtype, shape invariants, packed-length bounds, monotonic
    `cu_seqlens`, label-shift correctness, and per-segment loss-mask values
    by re-running `extract_segments` on each document inside a packed sample.

- **`tests/unit_tests/data/test_cp_utils.py`** *(232 lines)* — unit tests for
  `pad_thd_sequences_for_cp` in `megatron/core/utils.py`.
  - Compares against a slow obviously-correct `_reference_pad` implementation.
  - Covers: segments shorter than the divisibility factor, mixed lengths,
    already-divisible (no-op), 2-D input squeezing, dtype preservation
    (`int32`/`int64` for `cu_seqlens`, mixed dtypes per tensor), multi-tensor
    calls, single-tensor calls, `divisibility_factor=1`, and randomized
    parametric tests checking the post-condition that every padded segment
    length is a multiple of the divisibility factor.

### Updated tests

- **`tests/unit_tests/data/test_get_batch.py`** — the mock SFT data iterator
  now matches the new dataset's output shape (un-padded, no `position_ids`,
  no `max_seqlen`). Assertions across `test_sft_batch` and
  `test_hybrid_cp_batch` updated to expect `cu_seqlens_padded` to be
  populated when `cp_size > 1` (now produced by `preprocess_sft_batch`).

- **`tests/unit_tests/tokenizers/test_tokenizer.py`** — `test_null_tokenizer`
  and `test_detokenize_skip_special_tokens_unsupported_backend` updated to
  reflect the `NullTokenizer`'s new character-level encoding.

### Pre-checks

- [X] I have added relevant unit tests
- [X] I have added relevant functional tests
- [X] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [X] I have added relevant documentation
- [X] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Comment by asolergi-nv** [comment]
/ok to test [b1f4fc2](https://github.com/NVIDIA/Megatron-LM/pull/5017/commits/b1f4fc2781290881c43f6a37f632a52e73a430b0)

**Comment by asolergi-nv** [comment]
/ok to test [efa5da6](https://github.com/NVIDIA/Megatron-LM/pull/5017/commits/efa5da69e629d30e7ef754fa77032de68f512151)

**Comment by asolergi-nv** [comment]
/ok to test [fe2c20e](https://github.com/NVIDIA/Megatron-LM/pull/5017/commits/fe2c20e38b824a40309a2494e8e811e8a4368889)

File: megatron/core/tokenizers/text/libraries/huggingface_tokenizer.py [hunk 1]
```diff
@@ -259,7 +259,7 @@ def ids_to_tokens(self, ids: List[int]) -> List[str]:
         tokens = self.tokenizer.convert_ids_to_tokens(ids)
         return tokens
 
-    def text_to_ids(self, text: str) -> List[int]:
+    def text_to_ids(self, text: str, add_special_tokens: bool = True) -> List[int]:
```
Comment (dimapihtar) [inline megatron/core/tokenizers/text/libraries/huggingface_tokenizer.py:262]: why do we need to pass special tokens here? Looks like it's not being used by function

File: megatron/core/tokenizers/text/libraries/bytelevel_tokenizer.py [hunk 2]
```diff
@@ -70,7 +70,7 @@ def tokens_to_text(self, tokens):
         """
         return self.ids_to_text(tokens)
 
-    def text_to_ids(self, text):
+    def text_to_ids(self, text, add_special_tokens=True):
```
Comment (dimapihtar) [inline megatron/core/tokenizers/text/libraries/bytelevel_tokenizer.py:73]: the same question here

File: megatron/core/datasets/sft_dataset.py [hunk 3]
```diff
@@ -0,0 +1,826 @@
...
+        log_single_rank(logger, logging.DEBUG, f"\t> time elapsed: {t_end - t_beg:4f} seconds")
+
+        log_single_rank(
+            logger, logging.DEBUG, f"> total number of samples: {sample_index.shape[0] - 1}"
+        )
+
+        return document_index, sample_index, shuffle_index
+
+
+def _classify_items(
```
Comment (dimapihtar) [inline megatron/core/datasets/sft_dataset.py:465]: what about moving these extra functions to separate file like sft_utils.py?

File: megatron/core/datasets/sft_dataset.py [hunk 4]
```diff
@@ -0,0 +1,826 @@
...
+    think_start_str: str = "<think>"
+    think_end_str: str = "</think>"
+    tool_call_start_str: str = "<tool_call>\n"
+    tool_call_end_str: str = "</tool_call>\n"
+    tool_response_start_str: str = "<tool_response>\n"
+    tool_response_end_str: str = "</tool_response>\n"
+
+
+@dataclass
+class SFTDatasetConfig(GPTDatasetConfig):
```
Comment (dimapihtar) [inline megatron/core/datasets/sft_dataset.py:64]: do you think it makes sense to move all the configs to separate `sft_configs.py`?

File: megatron/training/datasets/sft_dataset.py [hunk 5]
```diff
@@ -1,201 +0,0 @@
-# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
```
Comment (yaoyu-33) [inline megatron/training/datasets/sft_dataset.py:1]: is this okay to remove? or should be put in legacy.

File: megatron/core/tokenizers/text/libraries/huggingface_tokenizer.py (hunk 1 above)
Comment (asolergi-nv) [reply megatron/core/tokenizers/text/libraries/huggingface_tokenizer.py:262]: I added it to all tokenizers since we are using this argument in the [parent class](https://github.com/NVIDIA/Megatron-LM/pull/5017/changes#diff-1d2114abd9fce06b03a2eef2653f796fe3e51427705dfcab16f11b6f7cb3a5b4R77)

**Comment by asolergi-nv** [comment]
/ok to test [97b2c55](https://github.com/NVIDIA/Megatron-LM/pull/5017/commits/97b2c55e4c373a072c36e79104bcdac1d77c022f)
